PWAWidgets/README.md

```diff
@@ -0,0 +1,975 @@
+* content-feed
+* content-item
+
+For productivity apps:
+
+* email-list
+* task-item
+* task-list
+
+For auth-requiring Widgets:
+
+* login-prompt
+
+[A selection of template samples](./template-samples/README.md), composed using Adaptive Cards, accompany this Explainer.
+
+### Data flow
+
+Widgets support user interaction through one or more [developer-defined `WidgetAction` objects](#Defining-a-WidgetAction), which are analogous to a [`NotificationAction`](https://notifications.spec.whatwg.org/#dictdef-notificationaction).
+
+Data flow in a Templated Widget is largely managed in two ways:
+
+1. Data flows from the Service Worker to a Widget instance as part of the [`widgets.updateInstance()`](#widgetsupdateinstance) and [`widgets.updateByTag()`](#widgetsupdatebytag) methods.
+2. Data (in the form of interaction) flows from a Widget to the associated PWA’s Service Worker via a [`WidgetEvent`](#widget-related-events).
+
+Here is an example of how this might look in the context of a Periodic Sync:
+
+<figure id="periodic-sync">
+
+![](media/sync.gif)
+
+</figure>
+
+This video shows the following steps:
+
+1. As part of a Periodic Sync, the Service Worker makes a `Request` to the host or some other endpoint.
+2. The `Response` comes back.
+3. As the Service Worker is aware of which widgets rely on that data, via the `WidgetDefinition` provided during [install](#dfn-install), the Service Worker can identify which widgets need updating. (This is internal logic and not shown in the video).
+3. The Service Worker takes that data — perhaps packaging it with other instructions — and uses [`widgets.updateInstance()`](#widgetsupdateinstance) (or [`widgets.updateByTag()`](#widgetsupdatebytag)) to update the specific widgets that make use of that data.
+
+To show a more complicated example, consider what should happen if certain Widgets depend on authentication and the user happens to log out in the PWA or a browser tab. The developers would need to track this and ensure the Service Worker is notified so it can replace any auth-requiring Widgets with a prompt back into the app to log in.
+
+Here’s how that might work:
+
+<figure id="user-logout">
+
+![](media/logout.gif)
+
+</figure>
+
+This video shows:
+
+1. The user logging out from the context of a `Client`. When that happens, the `Client`, sends a `postMessage()` to the Service Worker, alerting it to the state change in the app.
+2. The Service Worker maintains a list of active Widgets and is aware of which ones require authentication (informed by the `auth` property of the [`WidgetDefinition`](#defining-a-widget)). Knowing auth has been revoked, the Service Worker pushes a new template to each auth-requiring Widget with a notice and a button to prompt the user to log in again.
+
+The next step in this flow is for the user to log back in. They could do that directly in the Client, but let’s use the `WidgetAction` provided in the previous step:
+
+<figure id="user-login">
+
+![](media/login.gif)
```

Comment:
Alt text would be appropriate.

Reply:
The text below fully describes the image, which is why I didn’t include it.